core/src/main/java/io/grpc/Metadata.java

```diff
@@ -283,8 +395,25 @@ public boolean containsKey(Key<?> key) {
    */
   @ExperimentalApi
   public <T> void discardAll(Key<T> key) {
```

Comment:
I wonder if we can mege the `removeAll` and `discardAll` logic into one private method that takes a boolean argument `returnValues`.